Improve publication onboarding state synchronization.

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #9149 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on assets/js/modules/reader-revenue-manager/components/dashboard/PublicationApprovedOverlayNotification.js:
```diff
@@ -73,12 +88,16 @@ export default function PublicationApprovedOverlayNotification() {
+	 * - The publication onboarding state is complete.
 	 */
 	const shouldShowNotification =
 		isDismissed === false &&
 		showApprovedNotificationUI === true &&
 		! isViewOnly &&
-		dashboardType === VIEW_CONTEXT_MAIN_DASHBOARD;
+		dashboardType === VIEW_CONTEXT_MAIN_DASHBOARD &&
+		initialPublicationOnboardingStateChanged.current === true &&
+		publicationOnboardingState === ONBOARDING_COMPLETE;
```
Currently, the overlay notification no longer appears for me, because no one sets the value for `UI_KEY_READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION` in the `CORE_UI` store, which is only done when the sync is run on the client-side.

The logic here should be changed so that the notification is displayed when either the UI store value is set, OR, the new conditions are met.

Somewhat like the following:

```js
const shouldShowNotification =
	isDismissed === false &&
	! isViewOnly &&
	dashboardType === VIEW_CONTEXT_MAIN_DASHBOARD &&
	( showApprovedNotificationUI === true ||
		( initialPublicationOnboardingStateChanged.current === true &&
			publicationOnboardingState === ONBOARDING_COMPLETE ) );
```

Could you please investigate and ensure that the notification appears?

nfmohit reviewed (changes requested):
Brilliant work on this, thank you @ankitrox !

I have left a number of comments in the PR. The major issue that I've noticed is that the "publication approved overlay notification" does not appear when the cron event runs and the publication onboarding state changes to complete.

Could you please go through my comments and update the PR accordingly? Thank you!

ankitrox commented on includes/Modules/Reader_Revenue_Manager/Synchronize_OnboardingState.php:
```diff
@@ -0,0 +1,147 @@
+				return;
+			}
+
+			$publication = array_filter(
+				$publications,
+				function ( $pub ) use ( $publication_id ) {
+					return $pub->getPublicationId() === $publication_id;
+				}
+			);
+
+			// If publication is empty, return early.
+			if ( empty( $publication ) ) {
+				return;
+			}
+
+			$publication = reset( $publication );
```
I think we should keep it as is considering the edge case where we couldn't find the matching publication. In that case accessing [0] may lead to the warning.

nfmohit replied:
I understand your concern. I feel initialising the variable `$publication` with an array is a little odd, as it is a singular term but holds an array. What if we change the name of the variable to `$filtered_publications`? Also, what do you think about using `[0]` to access the first element of the array instead of using `reset`, as I think that makes the code further readable and doesn't bother with the internal pointer.

```suggestion
			$filtered_publications = array_filter(
				$publications,
				function ( $pub ) use ( $publication_id ) {
					return $pub->getPublicationId() === $publication_id;
				}
			);

			// If there are no filtered publications, return early.
			if ( empty( $filtered_publications ) ) {
				return;
			}

			$publication = $filtered_publications[0];

```

ankitrox commented on includes/Modules/Reader_Revenue_Manager/Synchronize_OnboardingState.php:
```diff
@@ -0,0 +1,147 @@
+		$connected              = $this->reader_revenue_manager->is_connected();
+		$cron_already_scheduled = wp_next_scheduled( self::CRON_SYNCHRONIZE_ONBOARDING_STATE );
+
+		if ( $connected && ! $cron_already_scheduled ) {
```
Yes, it makes sense.

nfmohit replied:
I think you have a point here. I retract my statement. I think we should revert to the original logic here. Thank you for pointing this out, Ankit!

nfmohit replied:
I think you have a point here. I retract my statement. I think we should revert to the original logic here. Thank you for pointing this out, Ankit!

ankitrox commented:
Thank you @nfmohit for reviewing the PR and adding the feedback. I've addressed the feedback and left response for couple of comments.

Over to you for another round of review.

Thanks again.

nfmohit commented on assets/js/modules/reader-revenue-manager/components/dashboard/PublicationApprovedOverlayNotification.js:
```diff
@@ -124,11 +130,21 @@ export default function PublicationApprovedOverlayNotification() {
 		}
-	}, [ saveSettings, setPublicationOnboardingStateChanged ] );
+	}, [
+		hasSettingsResolved,
```
I see `hasSettingsResolved` is a `useEffect` dependency but I don't see that used within the callback. Is this intentional? Do we only want to process the callback only when `hasSettingsResolved` is truthy?

ankitrox replied:
Thanks @nfmohit , yes, we need to run this effect when settings have been resolved. I've updated the condition inside `useEffect`.

nfmohit commented on assets/js/modules/reader-revenue-manager/components/dashboard/PublicationApprovedOverlayNotification.js:
```diff
@@ -61,10 +61,16 @@ export default function PublicationApprovedOverlayNotification() {
-	const initialPublicationOnboardingStateChanged = useRef(
-		publicationOnboardingStateChanged
+
+	// Check if settings has been resolved.
+	const hasSettingsResolved = useSelect( ( select ) =>
+		select( MODULES_READER_REVENUE_MANAGER ).hasFinishedResolution(
+			'getSettings'
+		)
```
Extremely nitpicky, but I believe `hasResolvedSettings` sounds more appropriate in this context, as otherwise, we'd want to change it to `haveSettingsResolved` as "settings" is "plural". `hasResolvedSettings` is also resonant to other similar variables in the codebase.

Additionally, I don't think the comment is necessary here as it is quite self-explanatory.

```suggestion
	const hasResolvedSettings = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).hasFinishedResolution(
			'getSettings'
		)
```

nfmohit commented on assets/js/modules/reader-revenue-manager/components/dashboard/PublicationApprovedOverlayNotification.js:
```diff
@@ -61,10 +61,16 @@ export default function PublicationApprovedOverlayNotification() {
+		)
 	);
 
+	const initialPublicationOnboardingStateChanged = useRef( undefined );
```
Is it necessary to specify `undefined` here? The initial value of `useRef` is `undefined` by default.

```suggestion
	const initialPublicationOnboardingStateChanged = useRef();
```

nfmohit reviewed (changes requested):
Thank you for making the updates, @ankitrox!

I have left a few more minor comments. We should be able to get this merged as soon as they are addressed. Thank you for your patience!

ankitrox commented:
Thanks @nfmohit for reviewing this. Sorry for missing couple of points from previous one, did not notice that there was a response added.

I have made the changes requested by you and sending it your way again for a review.

Thanks.

nfmohit commented on includes/Modules/Reader_Revenue_Manager/Synchronize_OnboardingState.php:
```diff
@@ -0,0 +1,161 @@
+				return;
+			}
+
+			$filtered_publications = array_filter(
+				$publications,
+				function ( $pub ) use ( $publication_id ) {
+					return $pub->getPublicationId() === $publication_id;
+				}
+			);
+
+			// If there are no filtered publications, return early.
+			if ( empty( $filtered_publications ) ) {
+				return;
+			}
+
+			$publication = $filtered_publications[0];
```
Okay, this is a bit embarrassing, but when I went to test this, I got an `Undefined array key 0` PHP warning and the synchronization certainly didn't work.

I did some digging and it looks like `array_filter` preserves the original array keys. When we access the first element using `[0]`, it may not exist if the original array keys are not sequential.

We either have to use `array_values` to re-index the filtered array (see suggestion below), or revert to using your original `reset` function (sorry!).

```suggestion
			$filtered_publications = array_filter(
				$publications,
				function ( $pub ) use ( $publication_id ) {
					return $pub->getPublicationId() === $publication_id;
				}
			);

			// If there are no filtered publications, return early.
			if ( empty( $filtered_publications ) ) {
				return;
			}

			// Re-index the filtered array to ensure sequential keys.
			$filtered_publications = array_values( $filtered_publications );

			$publication = $filtered_publications[0];
```

nfmohit reviewed (changes requested):
Thanks so much, @ankitrox, for your work on this. I've left my last couple comments on this one, after which I hope we can merge this. Thank you again for your patience!

ankitrox commented on assets/js/modules/reader-revenue-manager/index.js:
```diff
@@ -61,7 +59,7 @@ export const registerModule = ( modules ) => {
 			const homeURL = registry.select( CORE_SITE ).getHomeURL();
 
-			if ( isURLUsingHTTPS( homeURL ) ) {
+			if ( ! isURLUsingHTTPS( homeURL ) ) {
```
This is committed by mistake @nfmohit .

I was testing locally without https, thus made this change temporarily.

nfmohit replied:
Ah, that makes sense, no problem, thank you for confirming!

ankitrox commented:
Thanks @nfmohit .

I've made the changes and assigned to you to merge this.
